services/s3/generated.go

```diff
@@ -82,269 +73,289 @@ func GetStorageSystemMetadata(s *StorageMeta) StorageSystemMetadata {
-var _ Servicer = &Service{}
-
-type ServiceFeatures struct {
-}
-
-// pairServiceNew is the parsed struct
-type pairServiceNew struct {
-	pairs []Pair
-
-	// Required pairs
-	HasCredential bool
-	Credential    string
-	// Optional pairs
-	HasDefaultServicePairs bool
-	DefaultServicePairs    DefaultServicePairs
-	HasEndpoint            bool
-	Endpoint               string
-	HasForcePathStyle      bool
-	ForcePathStyle         bool
-	HasHTTPClientOptions   bool
-	HTTPClientOptions      *httpclient.Options
-	HasServiceFeatures     bool
-	ServiceFeatures        ServiceFeatures
-	HasUseAccelerate       bool
-	UseAccelerate          bool
-	HasUseArnRegion        bool
-	UseArnRegion           bool
-	// Enable features
-}
-
-// parsePairServiceNew will parse Pair slice into *pairServiceNew
-func parsePairServiceNew(opts []Pair) (pairServiceNew, error) {
-	result :=
-		pairServiceNew{pairs: opts}
+func WithUseArnRegion() types.Pair {
+	return types.Pair{Key: "use_arn_region", Value: true}
+}
 
-	for _, v := range opts {
+type Factory struct {
+	Credential          string
+	DefaultStorageClass string
+	EnableVirtualDir    bool
+	EnableVirtualLink   bool
+	Endpoint            string
+	ForcePathStyle      bool
+	Location            string
+	Name                string
+	UseAccelerate       bool
+	UseArnRegion        bool
+	WorkDir             string
+}
+
+func (f *Factory) FromString(conn string) (err error) {
+	slash := strings.IndexByte(conn, '/')
+	question := strings.IndexByte(conn, '?')
+
+	var partService, partStorage, partParams string
+
+	if question != -1 {
+		if len(conn) > question {
+			partParams = conn[question+1:]
+		}
+		conn = conn[:question]
+	}
+
+	if slash != -1 {
+		partService = conn[:slash]
+		partStorage = conn[slash:]
+	} else {
+		partService = conn
+	}
+
+	if partService != "" {
+		at := strings.IndexByte(partService, '@')
+		if at == -1 {
+			f.Endpoint = partService
+		} else {
+			xs := strings.SplitN(partService, "@", 2)
+			f.Credential, f.Endpoint = xs[0], xs[1]
+		}
+	}
+	if partStorage != "" {
+		slash := strings.IndexByte(partStorage[1:], '/')
+		if slash == -1 {
+			f.Name = partStorage[1:]
+		} else {
+			f.Name, f.WorkDir = partStorage[1:slash], partStorage[slash:]
```

Comment:
I'm not sure the connection string format. `s3://hmac:ak:sk@http:xxx:xxx/bucket/dir/?enable_virtual_dir` and `s3://bucket/dir/?endpoint=http:xxx:xxx&credential=hmac:ak:sk&enable_virtual_dir` both failed for `slice bounds out of range [1:0]` here.

Reply:
Oh, the input connection should trim the `s3://` prefix. I will add unit test and docs for this in next PR.

Reply:
Fixed in https://github.com/beyondstorage/go-storage/pull/1014